This function types

sandersn opened the pull request:
Implements the proposal at #6018 and finishes the work described at #3694. See the bottom of this description for a tutorial and usage recommendations for this new feature.

This change adds checking for this types in functions. With this-function types, you can prevent the use of methods as free functions. And this types allow you to make constructor functions that return a real type rather than `any`. A number of other common javascript patterns can be given types or even inferred with `--strictThis` turned on.

## Correctly prevent references to `this` when assigning callbacks

```ts
interface Callbacks {
  property: number;
  onClick(this: void, e: Event): void;
}
function handleClick(this:void, e: Event) {
  console.log(this.property); // error, 'void' has no member 'property'
}
class C {
  property: string
  tryHandleClick(this: this, e: Event): void {
    console.log(this.property); // OK, 'C' has member 'property'
  }
  actuallyHandleClick(this: void, e: Event): void {
    console.log(this.property); // error, 'void' has no member 'property'
  }
}
let c: C;
let callbacks: Callbacks;
callbacks.onClick = handleClick; // OK, this: void for both
callbacks.onClick = c.tryHandleClick; // Error, 'C' is not assignable to 'void'
callbacks.onClick = c.actuallyHandleClick; // OK, this: void for both
```

Note that callback functions are prevented from referring to members of `this` because they declare that `this: void` in order to be assignable to `onClick`.

Also note that the less common case of callback *methods* still allows *functions* to be assigned to these properties:

```ts
class Callbacks {
  m: number;
  callback: (this: this, n: number) => number;
}
function f(this: void, n: number) {
  return n;
}
let c: Callbacks;
c.callback = f; // OK, because f does not refer to any properties of `this`.
```

You can even require a *different* `this` for methods that will be used as a callback:

```ts
interface Callbacks {
  property: number;
  onClick(this: Callbacks, e: Event): void;
}
class C {
  property: string
  tryHandleClick(this: Callbacks, e: Event): void {
    console.log(this.property); // OK, 'C' has member 'property'
  }
}
let c: C;
let callbacks: Callbacks;
callbacks.onClick = c.tryHandleClick;
```

## Contextual typing of methods and functions in object literals

Now when an object literal declares that it is of some type, `this` is also contextually typed inside the object literal.

```ts
interface I {
  n: number;
  method(m: number): number;
  callback: (m: number) => number;
}
let o: I = {
  n: 12,
  method(m) {
    return this.n + m; // OK, `this: I` from context
  },
  callback = m => m + 1, // OK, `this: void` from context
}
```

## Build object literals from existing functions
You can also define functions alone and then later build an object literal from them. If the functions specify the `this` type, then the compiler will check references to this in the function's body. And it will also check that the function's `this` is assignable to the `this` of the object literal's type.

```ts
interface I {
  n: number;
  method: (m: number) => number;
}
function futureMethod(this: I, m: number) {
  return this.n + m;
}
let o: I = { n: 12, method: futureMethod };
```

Defining a function this way also requires that it cannot be called free:

```ts
futureMethod(12); // error, 'void' is not assignable to 'I'
o.futureMethod(12); // ok, 'o' is of type 'I'
```

## The type of functions when used as constructors is now known.

When you use `new` with a function, you can now declare the type of the object that will be constructed with `this`. Previously, the type was always `any`. The compiler will also check that your assignments to `this` are correct in the body of the function.

```ts
function Symbol(this: Symbol, flags: SymbolFlags, name: string) {
  this.flags = flags;
  this.name = name;
  this.declarations = undefined;
}
let s = new Symbol(flags, "core"); // s: Symbol
```

# `strictThisChecks` flag default types

## UPDATE: --strictThisChecks removed

Because of performance concerns and lack of information on how people will use `this` types, we decided to remove `--strictThisChecks` for this release. I left the original explanation below.

## --noImplicitThis

`--noImplicitThis` makes it an error to use `this` that is implicitly `any` inside of a function:

```ts
function F(x: number, y: number) {
    this.x = x; // this: any, so anything is legal
    this.z = y; // missed horrible typo
}
let f = new F(); // also, f: any  <-- that's sad :(
```

Add an annotation to fix this:

```ts
interface F { 
    x: number;
    y: number 
}
function F(this: F, x: number, y: number) {
    this.x = x;
    this.y = y;
}
let f = new F(12, 13); // f: F <-- hooray! :)
```

## Previous explanation of `strictThisChecks`

For backward compatibility, `this` types default to `any` if not specified. However, if `--strictThis` is specified, then functions will default `this: void` and methods will default `this: this`. This removes the need for most of the annotations in the previous examples. For example, an interface can declare a function with `this: void` by using function syntax and a method with `this: this` by using method syntax:

```ts
interface I {
  f: (n: number) => number; // this: void
  m(n: number): number; // this: this
}
function g(n: number) { // this: void
  return n;
}
class C {
  private special: number = 12;
  m(n: number): number { return n + this.special; }
}

let i: I;
let c: C;
i.f = g; // ok, this: void
i.f = c.m; // error, 'void' is not assignable to 'C' (missing member 'm')
i.m = c.m; // error, 'I' is not assignable to 'C' (missing member 'special')
```

This breaks a lot of existing code, but is easy to write for future code.

# How to upgrade code to `--strictThis`

When you switch on `--strictThis` you'll see a lot of errors with calling a method as if it were a function:

```ts
interface Object {
  method(n: number): void;
}
const f = object.method; 
f(12); // ERROR, this should be 'Object' not 'void'
```

You can fix the usage or the interface definition:

```ts
// Fix usage:
const fix1 = n => object.method(n);
correct(12); // OK, lambda captures the object and doesn't require 'this'
f.call(object, 12); // alternate fix: use Function.call

/// OR ///
// Fix definition:
interface Object {
  method(this: void, n: number): void;
}
// alternate fix: use function syntax to implicitly set this: void
interface Object {
  method: (n: number) => void;
}
const f = object.method;
f(12); // OK, this is void
```

# Designing new, strict-this code

To be able to switch `--strictThis` on, the style that you write interfaces needs to change. You need to consider how people will use your interfaces. If they will treat your interface's functions as methods on an object, you should use the normal method declaration syntax. If they will treat them as a callback or some other kind of free function, you should declare them using the function property syntax. For example:

```ts
interface Extractor {
  extract(input: string): Row[]; // method style
}

interface Callbacks {
  callback: (e: Event) => void; // function property style
}
```

A safe default is to use the method declaration syntax. I discuss the tradeoffs below.

## OO programming style
If you are writing pure OO Typescript/ES6, then you don't need to change much to work with  `--strictThis`. 

* OO interfaces should have `this: this`
* The only exception should be callbacks, which should have `this: void`.

Fortunately, these are the default types for the method and function syntax, respectively, so you don't have to write anything much different:

```ts
interface Extractor {
    extract(input: string): Row[];
    // the default (this: this) makes this the same as writing:
    // extract(this: this, input: string): Item[];

}
class XmlExtractor implements Extractor {
    extract(input: string): Row[] {
        // read from Xml into your Row objects
        // you can call private methods, etc, like before.
    }
    private helperMethod(sub: string): Row {
        // ...
    }
}
```

Now you can't assign `extract` to a function by mistake:

```ts
const xml = new XmlExtractor();
otherObject.onCall = xml.extract; // error: types of this are not compatible
let ex = xml.extract; // ok ...
ex('<row>...</row>'); // error, this: void is not compatible with this: XmlExtractor
otherObject.onExtract = ex; // error this: void is not compatible with this: XmlExtractor
```

And if you implement an interface, your methods get the right this-type regardless of what syntax you use:

```ts
interface Callbacks {
  callback: (e: Event) => void;
}
class XmlExtractor implements Extractor, Callbacks {
  extract(input: string): Row[] {
    // here, this: XmlExtractor
  }
  callback(e: Event) {
    // here, this: void
    this.extract(e.data); // error! 'this: void' has no method 'extract'
  }
}
otherObject.onExtract = xml.extract // error
otherObject.onExtract = xml.callback // ok!
```

### How can I use a method as a callback?

You may have noticed that `XmlExtractor.callback` isn't that useful as a method since you can't actually refer to any other methods. The solution is the same as you use today: wrap the method call inside a lambda:

```ts
class XmlExtractor implements Extractors, Callbacks {
  extract(input: string): Row[] { ... }
  callback = e => this.extract(e.data); // ok, => doesn't capture 'this'
}
```

This formulation is OK because lambda (`=>`) doesn't bind `this`, so `this` comes from the class instead of from the implementing function. 

## Functional style

If you are writing your code in functional style, you probably use interfaces to describe records of functions. You can still declare the functions using the method style and build instances of the interface using an object literal of your functions:

```ts
interface Compiler {
  parse(program: string): Node;
  bind(tree: Node): Map<Symbol>;
  check(tree: Node): Diagnostic[];
}
function parse(program: string) {
  // code inside does not refer to this
}
// etc ...
let compiler: Compiler = { parse, bind, check }
```

But this prevents users from pulling these functions off of `compiler` and using them individually:

```ts
let tree = compiler.parse("console.log('hello')");
let parse = compiler.parse;
parse("console.log('goodbye')"); // error, this is 'void' which is not assignable to 'Compiler'
```

This is safe because of course they *are* free functions that do not require access to this. If you want enable this usage, you need to use the function property declaration syntax for the interface:

```ts
interface Compiler {
  parse: (program: string) => Node;
  bind: (tree: Node) => Map<Symbol>;
  check: (tree: Node) => Diagnostic[];
}
function parse(program: string) {
  // code inside doesn't refer to this
}
let compiler: Compiler = { parse, bind, check }
let parse = compiler.parse;
parse("console.log('goodbye')");
```

### How can I implement an OO-style interface with nothing but functions?

If you want to implement an OO-style interface using only functions, you'll need to declare the `this` type in order to have access to that interface's members inside the function body.

```ts
function extract(this: Extractor, input: string): Row[] {
  // code here can refer to members of Extractor
}
```

Then you can create an object literal that contains that function:

```ts
let extractor: Extractor = {
  extract
}
```

This is essentially the code you would write today, but now usages of `this` are checked. For even more convenience you can use the contextual typing that comes from writing a function inside an object literal:

```ts
let extract: Extractor = {
  function extract(input: string): Row[] {
    // code here can refer to members of Extractor
  }
}
```

But at this point you might as well write a new class that implements Extractor.

## Interfacing with JavaScript

The problem with interfacing to JavaScript is that it may use any or all of the above styles. A good default for writing types for Javascript code is the method declaration style. This makes implementing the interface easier. On the other hand, it prevents users of the interface from pulling functions off of the interface in order to save them as variables or to use them as callbacks. 

This is the main reason that DefinitelyTyped definitions can't automatically compiled with --strictThis: it's impossible to predict which style of usage is the desired one for a currently-unmarked interface:

```ts
interface WholeClass {
  method(): void; // probably a method?
}
interface Record {
  func(x: string): number; // probably a function?
}
```

sandersn commented:
@ahejlsberg and @DanielRosenwasser, I believe you were both interested in this.

ahejlsberg commented on src/compiler/binder.ts:
```diff
@@ -1297,6 +1297,9 @@ namespace ts {
                     // name.
+                    if (options.strictThis) {
+                        seenThisKeyword = true;
+                    }
```
So any interface containing a method is now generic. Probably unavoidable, but we should get some data on what it means for performance.

sandersn replied:
I ran some numbers this afternoon and didn't see a big change. Could be misreading the results though.

ahejlsberg replied:
Just to recap your findings: On the Monaco project this ends up adding as much as 10% to the check time for an overall impact of up to 5%.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -2194,10 +2194,17 @@ namespace ts {
-            function buildDisplayForParametersAndDelimiters(parameters: Symbol[], writer: SymbolWriter, enclosingDeclaration?: Node, flags?: TypeFormatFlags, symbolStack?: Symbol[]) {
+            function buildDisplayForParametersAndDelimiters(thisType: Type, parameters: Symbol[], writer: SymbolWriter, enclosingDeclaration?: Node, flags?: TypeFormatFlags, symbolStack?: Symbol[]) {
                 writePunctuation(writer, SyntaxKind.OpenParenToken);
+                const useThisType = thisType && thisType.symbol;
```
Why the check for `thisType.symbol`?

sandersn replied:
`voidType` doesn't have a symbol, which avoids displaying (this: void, ...) for every single function, especially those that weren't annotated in the first place.

ahejlsberg replied:
There are other types that don't have a symbol, e.g. union and intersection types. If you want to exclude the `void` type then you should check for that.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -3525,12 +3533,13 @@ namespace ts {
         }
 
-        function createSignature(declaration: SignatureDeclaration, typeParameters: TypeParameter[], parameters: Symbol[],
+        function createSignature(declaration: SignatureDeclaration, typeParameters: TypeParameter[], parameters: Symbol[], thisType: Type,
```
Shouldn't `thisType` come before `parameters`?

sandersn replied:
fixed

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -4092,7 +4106,15 @@ namespace ts {
                         paramSymbol = resolvedSymbol;
                     }
-                    parameters.push(paramSymbol);
+                    if (paramSymbol.name === "this") {
```
Parser allows `this` parameters in any position, so here I would change to `if (i === firstParamIndex && paramSymbol.name === "this")` and then add `const firstParamIndex = isJSConstructSignature ? 1: 0` above the for loop.

ahejlsberg replied:
Actually, since we don't care about JSConstructSignature case, you should just say `if (i === 0 && paramSymbol.name === "this")`.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -4092,7 +4106,15 @@ namespace ts {
-                    parameters.push(paramSymbol);
+                    if (paramSymbol.name === "this") {
+                        thisType = param.type && getTypeOfSymbol(paramSymbol);
+                        if (i !== 0 || declaration.kind === SyntaxKind.Constructor) {
```
This check seems out of place here. It should be done in `checkParameter`.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -4092,7 +4106,15 @@ namespace ts {
+                    if (paramSymbol.name === "this") {
+                        thisType = param.type && getTypeOfSymbol(paramSymbol);
+                        if (i !== 0 || declaration.kind === SyntaxKind.Constructor) {
+                            error(param, Diagnostics.this_cannot_be_referenced_in_current_location);
```
I would have one error message that says `'this' parameter must be the first parameter` and another that says `A constructor cannot have a 'this' parameter`.

sandersn replied:
done

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -4092,7 +4106,15 @@ namespace ts {
                     }
-                    parameters.push(paramSymbol);
+                    if (paramSymbol.name === "this") {
+                        thisType = param.type && getTypeOfSymbol(paramSymbol);
```
So, if the programmer omits a type annotation on a `this` parameter, you will think there was *no* `this` parameter and get your parameter count adjustments wrong below. You need to track the presence of a this parameter separately from its type. When the type annotation is missing you should use `unknownType`. Also, you can just use `getTypeFromTypeNode(param.type)` instead of `getTypeOfSymbol`.

sandersn replied:
done

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -4100,7 +4122,7 @@ namespace ts {
                     if (param.initializer || param.questionToken || param.dotDotDotToken) {
                         if (minArgumentCount < 0) {
-                            minArgumentCount = i;
+                            minArgumentCount = i - (thisType ? 1 : 0);
```
See my comments about parameter count adjustments above.

sandersn replied:
done

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -4110,7 +4132,22 @@ namespace ts {
+                    else if ((declaration.kind === SyntaxKind.MethodDeclaration || declaration.kind === SyntaxKind.MethodSignature)
+                        && (isClassLike(declaration.parent) || declaration.parent.kind === SyntaxKind.InterfaceDeclaration)) {
+                        thisType = declaration.flags & NodeFlags.Static ?
+                            getWidenedType(checkExpression((<ClassLikeDeclaration>declaration.parent).name)) :
```
Yikes, this `checkExpression` call looks all wrong. The `name` of a class declaration isn't an expression. You need to use `getTypeOfSymbol(getSymbolOfNode(declaration.parent))`.

ahejlsberg replied:
Also, why are you widening the type?

sandersn replied:
I tried `getTypeOfSymbol(getSymbolOfNode(...` first. According to my notes, I had trouble getting `typeof C` vs `C` in the static case.

But it looks like I was wrong. It works now. I'm pretty sure some other change I made was wrong at the time.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -4117,7 +4139,22 @@ namespace ts {
+                }
+                if (!thisType && compilerOptions.strictThis) {
+                    if (declaration.kind === SyntaxKind.FunctionDeclaration
+                        || declaration.kind === SyntaxKind.CallSignature
```
Elsewhere we don't use this style of putting `||` at the beginning of the line.

sandersn replied:
oops, it's a habit left over from non-ASI languages. Fixed.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -5182,7 +5220,14 @@ namespace ts {
 
         function isContextSensitiveFunctionLikeDeclaration(node: FunctionLikeDeclaration) {
-            return !node.typeParameters && node.parameters.length && !forEach(node.parameters, p => p.type);
+            if (compilerOptions.strictThis) {
```
Why this check? It doesn't matter if were in strict-this mode, we *always* want to ignore `this` parameters for purposes of this function.

sandersn replied:
I thought that in loose-this mode, a function with all arg types specified and no `this` is not contextually typed. But thinking about it a second time, I don't think that's true:

```ts
interface I {
  y: number;
  m(this: I, x: number);
}
let o: I = {
  m(x: number) {
    return x + this./**/
  }
}
```

It would in fact be nice if we would contextually type `this` so that we could offer `y` as a completion. I'll get rid of everything except the updated check.

sandersn replied:
done

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -5182,7 +5220,14 @@ namespace ts {
-            return !node.typeParameters && node.parameters.length && !forEach(node.parameters, p => p.type);
+            if (compilerOptions.strictThis) {
+                return !node.typeParameters &&
+                    (!forEach(node.parameters, p => p.type)
```
Maybe add a `hasParameterWithTypeAnnotation` function and do the appropriate compensation for `this` parameters in there.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -5259,6 +5304,22 @@ namespace ts {
+                const t = target.thisType || anyType;
+                if (s !== voidType) {
+                    // void sources are assignable to anything.
+                    let related = compareTypes(getApparentType(t), getApparentType(s), reportErrors);
```
Why are you comparing apparent types?

sandersn replied:
Same reason I used `getWidenedType` elsewhere. I'll get rid of the calls since they are wrong.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -3546,15 +3555,19 @@ namespace ts {
         }
 
+        function getParameterTypeAtIndex(signature: Signature, i: number, max: number, outOfRangeType?: Type): Type {
+            return i < max ? getTypeOfSymbol(signature.parameters[i]) : (outOfRangeType || getRestTypeOfSignature(signature));
```
I find this function and its uses hard to reason about. Why does the `outOfRangeType` take precedence over the rest signature type? Why do we need this *and* `getTypeAtPosition`?

sandersn replied:
switched to getTypeAtPosition

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -5069,6 +5106,7 @@ namespace ts {
             }
             const result = createSignature(signature.declaration, freshTypeParameters,
                 instantiateList(signature.parameters, mapper, instantiateSymbol),
+                signature.thisType ? instantiateType(signature.thisType, mapper) : undefined,
```
`signature.thisType && instantiateType(signature.thisType, mapper)`

sandersn replied:
done

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -6436,9 +6497,7 @@ namespace ts {
-                const s = i < sourceMax ? getTypeOfSymbol(source.parameters[i]) : getRestTypeOfSignature(source);
-                const t = i < targetMax ? getTypeOfSymbol(target.parameters[i]) : getRestTypeOfSignature(target);
-                callback(s, t);
+                callback(getParameterTypeAtIndex(source, i, sourceMax), getParameterTypeAtIndex(target, i, targetMax));
```
Why did this change?

sandersn replied:
it repeats the same code as in `getTypeAsPosition`. I just switched it to `getTypeAtPosition` and the whole function got simpler.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -5259,6 +5304,22 @@ namespace ts {
             target = getErasedSignature(target);
 
             let result = Ternary.True;
+            if (source.thisType || target.thisType) {
```
Shouldn't this be `source.thisType && target.thisType`?

sandersn replied:
done

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -7336,7 +7395,12 @@ namespace ts {
             }
-
+            if (isFunctionLike(container)) {
+                const signature = getSignatureFromDeclaration(container);
```
This isn't going to work when the containing function has a contextually typed `this` because the signature will appear to have no `this` type. I think you need to `getTypeOfSymbol(getSymbolOfNode(container))` and then first the `this` type out of the signatures of the type.

ahejlsberg replied:
Actually, `getTypeOfSymbol` won't work either. Actually, I'm unclear on when you contextually type `this`? Your writeup for the PR states that you contextually type `this` in methods in object literals, but it doesn't seem to work. In fact, I haven't been able to see any contextual typing of `this` in snippets I have written. We should discuss our strategy here.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -10305,11 +10400,17 @@ namespace ts {
+                    // do not contextually type thisType for ArrowFunction. 
+                    signature.thisType = context.thisType;
+                }
+            }
```
I haven't been able to write an example where a contextual `this` type actually gets assigned.

sandersn replied:
There's a test for it on line 164 and following:
https://github.com/Microsoft/TypeScript/pull/6739/files#diff-f0430cf5aca89cd84e8c29f53dbecbb4R164

Briefly,

```ts
interface I { a: number; m(): number }
let impl: I = { a: 12, m() { return this.a } }
```

ahejlsberg replied:
I think the problem is that your implementation doesn't work in "pull mode". Try pasting that example into VS Code and hover over `this`. You'll see type `any`. This is because something needs to have "pushed" a this type into `m`'s signature before you can pull it out in a check of `this`.

sandersn replied:
`--strictThis` needs to be on too, or the interface needs to be `interface I { a: number, m(this: this): number }`.

I haven't tried running the language service in an actual editor yet. Let me try that and see whether the explicit version works..

sandersn replied:
Being explicit doesn't work either. I was going off the .types result from tests, but I guess a full compile "pulls" everything in order?
 
Maybe you can show me how to switch from push to pull later today.

ahejlsberg replied:
A full compile generally has the property of checking containers before their children, but pull mode obviously doesn't do that. Yes, let's discuss later today.

sandersn replied:
ok, should be working now.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -9428,6 +9504,18 @@ namespace ts {
+                // If the source is not of the form `x.f`, then sourceType = voidType
+                // If the target is voidType, then the check is skipped -- anything is compatible.
+                // If the the expression is a new expression, then the check is skipped.
+                const calleeNode = node.kind === SyntaxKind.CallExpression && (<PropertyAccessExpression>(<CallExpression>node).expression).expression;
```
This check isn't safe. You need to check if `(node as CallExpression).expression.kind` is a property access or element access, and only then can you extract the target expression. I think you need a `getThisArgumentOfCall` helper that does the extraction.

sandersn replied:
done

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -9428,6 +9504,18 @@ namespace ts {
+                const calleeNode = node.kind === SyntaxKind.CallExpression && (<PropertyAccessExpression>(<CallExpression>node).expression).expression;
+                const calleeType: Type = calleeNode ? checkExpressionWithContextualType(calleeNode, signature.thisType, undefined) : voidType;
+                const errorNode = reportErrors ? (calleeNode || node) : undefined;
+                if (!checkTypeRelatedTo(calleeType, getApparentType(signature.thisType), relation, errorNode, headMessage)) {
```
What's with all of the apparent types?

sandersn replied:
Same confusion as before left over from trying to handle static this. Removed.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -9428,6 +9504,18 @@ namespace ts {
+                // If the target is voidType, then the check is skipped -- anything is compatible.
+                // If the the expression is a new expression, then the check is skipped.
+                const calleeNode = node.kind === SyntaxKind.CallExpression && (<PropertyAccessExpression>(<CallExpression>node).expression).expression;
+                const calleeType: Type = calleeNode ? checkExpressionWithContextualType(calleeNode, signature.thisType, undefined) : voidType;
```
The expression in the callee node has already been type checked by the time we get here (in a `x.f(...)` type of call we need to know the type of `x` before we can even figure out what `f` refers to), so it doesn't make sense to say that the callee node is contextually typed by the this type.

sandersn replied:
Removed.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -9801,8 +9888,13 @@ namespace ts {
+                const calleeNode = node.kind === SyntaxKind.CallExpression && (<PropertyAccessExpression>(<CallExpression>node).expression).expression;
+                if (calleeNode && isContextSensitive(calleeNode)) {
+                    excludeCallee = true;
+                }
```
I'm not sure what you're trying to accomplish here?

sandersn replied:
Expressions like  `(function (n) { return this.m + n }).call({ m: 101 }, 12)` should contextually give a type to `this.m`. But I guess this won't work if the function's type is already determined. I should look at my tests again, I thought I had one that showed this in action.

sandersn replied:
Nope. It fails. I'll take all the contextual typing code for this arguments out.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -9329,7 +9393,7 @@ namespace ts {
         }
 
-        function inferTypeArguments(node: CallLikeExpression, signature: Signature, args: Expression[], excludeArgument: boolean[], context: InferenceContext): void {
+        function inferTypeArguments(node: CallLikeExpression, signature: Signature, args: Expression[], excludeCallee: boolean, excludeArgument: boolean[], context: InferenceContext): void {
```
I don't see why you need this new parameter. The callee should always be handled the same in type inference, i.e. if the signature has a `thisType`, infer from the type of the callee expression to the `thisType`.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -9355,6 +9419,13 @@ namespace ts {
+                const calleeType: Type = calleeNode ? checkExpressionWithContextualType(calleeNode, signature.thisType, mapper) : voidType;
+                inferTypes(context, calleeType, signature.thisType);
+            }
+
```
See my comments for the corresponding code in `checkApplicableSignature`.

ahejlsberg commented on src/compiler/checker.ts:
```diff
@@ -9384,8 +9455,13 @@ namespace ts {
             if (excludeArgument) {
+                if (signature.thisType && calleeNode) {
+                    if (excludeCallee === false) {
+                        inferTypes(context, checkExpressionWithContextualType(calleeNode, signature.thisType, inferenceMapper), signature.thisType);
```
Just use `checkExpression(calleeNode)` here, don't supply a contextual type. Actually, we may want to consider caching that type so we don't have to do a full type check on the node again (we already did that in order to get here in the first place).

DanielRosenwasser commented on src/lib/core.d.ts:
```diff
@@ -215,14 +215,16 @@ interface Function {
       * @param argArray A list of arguments to be passed to the method.
       */
-    call(thisArg: any, ...argArray: any[]): any;
+    call<T,U>(this: (this: T, ...argArray: any[]) => U, thisArg: T, ...argArray: any[]): U;
```
Space after comma before `U` for `call` and `apply`

sandersn replied:
done

DanielRosenwasser commented on src/lib/core.d.ts:
```diff
@@ -230,7 +232,8 @@ interface Function {
       * @param argArray A list of arguments to be passed to the new function.
       */
-    bind(thisArg: any, ...argArray: any[]): any;
+    bind<T, U>(this: (this: T, ...argArray: any[]) => U, thisArg: T, ...argArray: any[]): (this: void, ...argArray: any[]) => U;
```
Wouldn't you want to place some constraint between `T` and `U`?

sandersn replied:
`T` is the function's `this` type. `U` is the function's return value. There's no relation between them. I don't know TypeScript's conventions for type variable names that well. Would it be easier to read to change this to `<TThis, TReturn>`?

DanielRosenwasser commented on src/lib/core.d.ts:
```diff
@@ -230,7 +232,8 @@ interface Function {
       */
-    bind(thisArg: any, ...argArray: any[]): any;
+    bind<T, U>(this: (this: T, ...argArray: any[]) => U, thisArg: T, ...argArray: any[]): (this: void, ...argArray: any[]) => U;
+    bind(this: Function, thisArg: any, ...argArray: any[]): any;
```
Is there a reason you need the explicit `this`-arg?

sandersn replied:
1. I would like `this` to show up in the quick info. It's quite relevant to this method.
2. I want to say `this: Function` instead of `this: this` because it's a smaller type (thus, easier to call, harder to implement). I *hope* you can't monkey patch `bind` so implementation of `bind` shouldn't be an issue ... I'll go see what I can learn about that. It seems like something people might actually do in javascript.
3. `this: Function` is easier to read than `this: this` for first-time readers of `this` parameters.
3. Uniformity with the first overload.

sandersn commented:
Trying to compile Angular 2 exposes bug #7097. Until that bug is fixed, `--strictThis` will cause high memory usage or out-of-memory crashes in projects with classes that have a lot of static, generic functions (Angular 2's ListWrapper class has 25 static functions in it).

Otherwise I've addressed all the comments so far. @ahejlsberg @DanielRosenwasser any other comments?

sandersn commented:
#7138 fixes #7097, which fixes the high memory/out-of-memory bug for classes with lots of static, generic functions when running `--strictThis`.

sandersn commented:
I updated the PR with two new features:

1. `this` expressions inside object literal methods now have the type of the containing object literal instead of `any`.
2. When unioning two call signatures, the this-type of the signature is now unioned if there is an otherwise-common call signature. This resolves the tricky error resulting from unioned call signatures having the wrong this-type. This applies to types with multiple call signatures, too.

sandersn commented on src/compiler/checker.ts:
```diff
@@ -8050,6 +8110,10 @@ namespace ts {
                 }
             }
 
+            if (compilerOptions.noImplicitThis && isFunctionLike(container)) {
```
should give an error in all cases, not just `isFunctionLike(container)`

mhegazy replied:
we should not need the `isFunctionLike` here. it should be an error to use `this` in the global scope with `noImplicitThis`

mhegazy replied:
we should add a test for that as well.

sandersn replied:
done. There was already a test -- but it asserted that it was not an error.

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -5357,7 +5378,12 @@ namespace ts {
+            if (node.kind === SyntaxKind.ArrowFunction) {
+                return !node.typeParameters && node.parameters.length && areAllParametersUntyped;
+            }
+            const hasThisType = node.parameters.length && (<Identifier>node.parameters[0].name).text === "this" && node.parameters[0].type;
```
this check is subsumed by the above check for areAllParametersUntyped. do we need it?

sandersn replied:
Turns out no. It still needs special-casing for arrow functions -- they should be skipped if they don't have arguments; nothing else should since everything else implicitly has `this`. But the function is much simpler now.

DanielRosenwasser replied:
What exactly do you mean? If I have a `this` parameter and it lacks a type, won't this be true if `areAllParametersUntyped`?

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -8020,12 +8063,29 @@ namespace ts {
                 const type = container.flags & NodeFlags.Static ? getTypeOfSymbol(symbol) : (<InterfaceType>getDeclaredTypeOfSymbol(symbol)).thisType;
                 return getNarrowedTypeOfReference(type, node);
             }
+            if (container.parent && container.parent.kind === SyntaxKind.ObjectLiteralExpression) {
```
i would move this up into `isFunctionLike` check above.

sandersn replied:
done

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -12285,6 +12397,14 @@ namespace ts {
+                if (indexOf(func.parameters, node) !== 0) {
+                    error(node, Diagnostics.this_parameter_must_be_the_first_parameter);
+                }
+                if (func.kind === SyntaxKind.Constructor) {
```
what about ConstructSignature?

mhegazy replied:
and ConstructorType, and SetAccessor

sandersn replied:
good catch. done.

mhegazy commented on src/compiler/commandLineParser.ts:
```diff
@@ -123,6 +123,10 @@ namespace ts {
         },
         {
+            name: "noImplicitThis",
+            type: "boolean",
```
Please add a discription for this one.

mhegazy commented:
can we add a fourslash test for parameter help.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -5442,6 +5468,22 @@ namespace ts {
+                        related = compareTypes(source.thisType, target.thisType, /*reportErrors*/ false);
+                        if (!related) {
+                            if (reportErrors) {
+                                errorReporter(Diagnostics.Types_of_parameters_0_and_1_are_incompatible, "this", "this");
```
Wow, this sounds confusing! Why not make a new error message like

> Types of each signature's 'this'-context are incompatible.

sandersn replied:
It's always the second line of an elaboration whose first line looks like

```ts
Type '(this: D, m: number) => number' is not assignable to type '(this: C, m: number) => number'.
```

Since `this` shows up in the signatures as the first parameter, I don't think there's enough value in a custom message here.
```

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -5442,6 +5468,22 @@ namespace ts {
 
             let result = Ternary.True;
+            if (source.thisType && target.thisType) {
+                if (source.thisType !== voidType) {
```
Consider folding this into the previous check to remove a level of nesting.

sandersn replied:
done

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -5442,6 +5468,22 @@ namespace ts {
+            if (source.thisType && target.thisType) {
+                if (source.thisType !== voidType) {
+                    // void sources are assignable to anything.
+                    let related = compareTypes(target.thisType, source.thisType, reportErrors);
```
Why not write this as below, and using the same direction on the relation?

```ts
const related = compareTypes(source.thisType, target.thisType, /*reportErrors*/ false) ||
    compareTypes(target.thisType, source.thisType, reportErrors);
if (!related) {
    if (reportErrors) {
        // ...
    }
    return related;
}
```

sandersn replied:
I exploded the code into this expanded form during development. Fixed.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -10119,6 +10202,18 @@ namespace ts {
         function checkApplicableSignature(node: CallLikeExpression, args: Expression[], signature: Signature, relation: Map<RelationComparisonResult>, excludeArgument: boolean[], reportErrors: boolean) {
+            const headMessage = Diagnostics.Argument_of_type_0_is_not_assignable_to_parameter_of_type_1;
+            if (signature.thisType && signature.thisType !== voidType && node.kind !== SyntaxKind.NewExpression) {
+                // If the source's this is not of the form `x.f` or `x[f]`, then sourceType = voidType
```
You mean `x["f"]`?

sandersn replied:
Yup

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -10119,6 +10202,18 @@ namespace ts {
         }
 
         function checkApplicableSignature(node: CallLikeExpression, args: Expression[], signature: Signature, relation: Map<RelationComparisonResult>, excludeArgument: boolean[], reportErrors: boolean) {
+            const headMessage = Diagnostics.Argument_of_type_0_is_not_assignable_to_parameter_of_type_1;
```
Isn't this going to be a confusing error message for users?

sandersn replied:
Yeah, it's not great. Ryan suggested the custom message

'this' context of type '{0}' is not assignable to method 'this' of type '{1}'

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -10119,6 +10202,18 @@ namespace ts {
+            const headMessage = Diagnostics.Argument_of_type_0_is_not_assignable_to_parameter_of_type_1;
+            if (signature.thisType && signature.thisType !== voidType && node.kind !== SyntaxKind.NewExpression) {
+                // If the source's this is not of the form `x.f` or `x[f]`, then sourceType = voidType
+                // If the target's this is voidType, then the check is skipped -- anything is compatible.
```
What is the `source` and what is the `target` here?

sandersn replied:
:+1: Nice rewording.

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -1875,7 +1875,23 @@
+        "category": "Error",
+        "code": 2678
+    },
+    "'this' parameter must be the first parameter.": {
```
A 'this' parameter

sandersn replied:
done

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -12419,11 +12414,14 @@ namespace ts {
+                if (func.kind === SyntaxKind.Constructor || func.kind === SyntaxKind.ConstructSignature) {
                     error(node, Diagnostics.A_constructor_cannot_have_a_this_parameter);
                 }
+                if (func.kind === SyntaxKind.SetAccessor) {
```
still missing constructorType.

`var x: new (this: number)=> number`

sandersn replied:
Are you looking at the latest commits? I added to the check on line 12419 and an associated test (with an interface, not a variable, but they fail the same way) in thisTypeInFunctionsNegative.ts

mhegazy replied:
you have construct signature covered, i.e. `{ new(this: T): T; }`, i am talking about `ConstructorType` i.e. `var x: new (this: number)=> number`. this still fails on the latest from this branch.

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -1879,6 +1879,30 @@
         "category": "Error",
         "code": 2678
     },
+    "A function that is called with the 'new' keyword cannot have a 'this' type that is void.": {
```
Void in single quotes

sandersn replied:
done

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -1879,6 +1879,30 @@
+        "category": "Error",
+        "code": 2683
+    },
+    "'this' context of type '{0}' is not assignable to method 'this' of type '{1}'.": {
```
method -> method's

sandersn replied:
done

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -5451,6 +5474,18 @@ namespace ts {
+                    || compareTypes(target.thisType, source.thisType, reportErrors);
+                if (!related) {
+                    if (reportErrors) {
+                        errorReporter(Diagnostics.Types_of_parameters_0_and_1_are_incompatible, "this", "this");
```
1. "The specified 'this' types of each signature are incompatible.
2. "The 'this'-context for each signature are incompatible."
3. "Types of each 'this' parameter are incompatible

I'm leaning towards 1.

sandersn replied:
I went with a cut-down (1) -- I got rid of "The specified"

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -1879,6 +1879,34 @@
+        "category": "Error",
+        "code": 2684
+    },
+    "'this' types of each signature are incompatible.": {
```
Every time I think about this audibly, I hear a garden path sentence for the first two words:

"This type's..."

sandersn replied:
After discussing this in person, we couldn't come up with anything better. Anyone else have any ideas? 

I'll plan to merge it as-is and change it later if anybody comes up with something better.

Arnavion replied:
Prepend `The`? And to the previous one as well?

`"'this' context of type '{0}' is not assignable to method's 'this' of type '{1}'."` seems to be written to avoid articles on purpose, but then there's `"A setter cannot have a 'this' parameter."`

sandersn replied:
Well, that's 3 people who like prepending "The" as a solution. I'll do that then.

Note that after much discussion yesterday, @DanielRosenwasser @RyanCavanaugh and I decided not to try for consistency of determiner usage in error messages just yet. The majority doesn't use them but here they seem to help so I think that outweighs the inconsistency.

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -12305,6 +12412,17 @@ namespace ts {
+                if (indexOf(func.parameters, node) !== 0) {
+                    error(node, Diagnostics.A_this_parameter_must_be_the_first_parameter);
+                }
+                if (func.kind === SyntaxKind.Constructor || func.kind === SyntaxKind.ConstructSignature) {
```
Re posting, since the other comment got removed. 

you have `ConstructSignature` covered, i.e. `{ new(this: T): T; }`, i am talking about `ConstructorType` i.e. `var x: new (this: number)=> number.` this still fails on the latest from this branch.

sandersn replied:
ok, got it. fixed.

mhegazy commented:
:+1:

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -4238,7 +4252,13 @@ namespace ts {
                         paramSymbol = resolvedSymbol;
                     }
-                    parameters.push(paramSymbol);
+                    if (i == 0 && paramSymbol.name === "this") {
```
`===`

sandersn replied:
nice catch

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -8040,7 +8076,24 @@ namespace ts {
+                if (container.parent && container.parent.kind === SyntaxKind.ObjectLiteralExpression) {
+                    // Note: this works because object literal methods are deferred,
+                    // which means that the type of the containing object literal is already known.
+                    const type = checkExpressionCached(<ObjectLiteralExpression>container.parent);
```
Huh - we should add a regression test to ensure that a set of mutually recursive object literal methods type-checks properly.

```ts
var x = {
    a: 100,

    start() {
        return this.doX(this.a);
    }    

    doX(n: number) {
        this.doY(n);
    }

    doY(n: number): number {
        if (n > 10) {
            return this.doX(n - 10);
        }
        if (n > 5) {
            return this.doY(n - 1);
        }
        return n;
    }
}

var i: number = x.start();
```

And maybe create an interface with the same shape, create a variable of that type, and assign `x` to it.

sandersn replied:
done

DanielRosenwasser commented:
It looks like the only test where you have a `this` parameter with no type annotation is on an arrow function. Can you write

1. A test with a contextually typed explicit `this` parameter.
2. A test with an implicit `any` on an explicit `this` parameter.

sandersn commented:
(2) is on the last line of thisTypeInFunctions.ts
I'll add (1).

Gaelan commented:
Not that #7097 is fixed, can strictThisChecks be readded?

DanielRosenwasser commented:
@Gaelan I believe we concluded at our design meeting that we were going to hold off on `--strictThis`, partially for perf reasons, partially because alongside `readonly` and non-nullable types, this was going to cause a lot of issues for the community in updating `.d.ts` files in a compatible manner.

Gaelan commented:
@DanielRosenwasser Ah.

sandersn commented:
@DanielRosenwasser I added a contextual typing test as in (2) with an implicit any on a this parameter. Note that the types are not *quite* right &mdash; the `this` parameter is never contextually typed, just `this` expressions in the function body. This is because `getSignatureFromDeclaration` stores `thisType` eagerly -- not `thisSymbol` like the rest of the compiler. I discussed this with @ahejlsberg and we decided it was OK for now since it simplifies the code a lot. I might change it later.

The effect is that this:

```ts
o.someMethod = function(this, m) { return this.n + m };
```

has these types:
```ts
o.someMethod: (this: O, m: number) => number // source of context
function(this: any, m: number) => number // this: any, but returns number
this: O // inside the body, this is contextually typed
this.n: number
m: number
return this.n +  m // correctly returns number
```

Note that the `thisType: any` is not contagious because it's not actually used -- for `this.n`, `checkThisExpression` uses the contextual type instead, which is `this: O`.

heruan commented:
If I understand correctly, with this PR the following should work:
```typescript
class Foo {
    private x: string;
}
class Bar {
    private foo: Foo;

    constructor(foo: Foo) {
        this.foo = foo;
    }

    actAsFoo(this: Foo): string {
        return this.x;
    }

    getFromFoo(): string {
        return this.actAsFoo.call(this.foo);
    }
}
```

butin `actAsFoo` I get `Error: Property 'x' is private and only accessible within class 'Foo'.

avivcarmis commented:
This feature is exactly what I'm after, I see it was merged but i can't find any docs regarding, and i saw here that it was not working well.
Is there any update? Can it be found in some newer version? Will it be added in the future?

sandersn commented:
https://www.typescriptlang.org/docs/handbook/functions.html has the documentation near the bottom of the page.
